score.go

```diff
@@ -0,0 +1,952 @@
+func (ps *peerScore) Leave(topic string) {}
+
+func (ps *peerScore) Graft(p peer.ID, topic string) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return
+	}
+
+	tstats, ok := pstats.getTopicStats(topic, ps.params)
+	if !ok {
+		return
+	}
+
+	tstats.inMesh = true
+	tstats.graftTime = time.Now()
+	tstats.meshTime = 0
+	tstats.meshMessageDeliveriesActive = false
+}
+
+func (ps *peerScore) Prune(p peer.ID, topic string) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return
+	}
+
+	tstats, ok := pstats.getTopicStats(topic, ps.params)
+	if !ok {
+		return
+	}
+
+	// sticky mesh delivery rate failure penalty
+	threshold := ps.params.Topics[topic].MeshMessageDeliveriesThreshold
+	if tstats.meshMessageDeliveriesActive && tstats.meshMessageDeliveries < threshold {
+		deficit := threshold - tstats.meshMessageDeliveries
+		tstats.meshFailurePenalty += deficit * deficit
+	}
+
+	tstats.inMesh = false
+}
+
+func (ps *peerScore) ValidateMessage(msg *Message) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	// the pubsub subsystem is beginning validation; create a record to track time in
+	// the validation pipeline with an accurate firstSeen time.
+	_ = ps.deliveries.getRecord(ps.msgID(msg.Message))
+}
+
+func (ps *peerScore) DeliverMessage(msg *Message) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	ps.markFirstMessageDelivery(msg.ReceivedFrom, msg)
+
+	drec := ps.deliveries.getRecord(ps.msgID(msg.Message))
+
+	// mark the message as valid and reward mesh peers that have already forwarded it to us
+	drec.status = delivery_valid
+	drec.validated = time.Now()
+	for p := range drec.peers {
+		// this check is to make sure a peer can't send us a message twice and get a double count
+		// if it is a first delivery.
+		if p != msg.ReceivedFrom {
+			ps.markDuplicateMessageDelivery(p, msg, time.Time{})
+		}
+	}
+}
+
+func (ps *peerScore) RejectMessage(msg *Message, reason string) {
+	ps.Lock()
+	defer ps.Unlock()
+
+	// TODO: the reasons should become named strings; good enough for now.
+	switch reason {
+	// we don't track those messages, but we penalize the peer as they are clearly invalid
+	case "missing signature":
+		fallthrough
+	case "invalid signature":
+		ps.markInvalidMessageDelivery(msg.ReceivedFrom, msg)
+		return
+
+		// we ignore those messages, so do nothing.
+	case "blacklisted peer":
+		fallthrough
+	case "blacklisted source":
+		return
+
+	case "validation queue full":
+		// the message was rejected before it entered the validation pipeline;
+		// we don't know if this message has a valid signature, and thus we also don't know if
+		// it has a valid message ID; all we can do is ignore it.
+		return
+	}
+
+	drec := ps.deliveries.getRecord(ps.msgID(msg.Message))
+
+	if reason == "validation throttled" {
+		// if we reject with "validation throttled" we don't penalize the peer(s) that forward it
+		// because we don't know if it was valid.
+		drec.status = delivery_throttled
```

Comment:
nit: non-idiomatic go variable name.

Reply:
it's an unexported constant.

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

Reply:
sigh... done.